addons/web/static/src/webclient/company_service.js

```diff
@@ -51,9 +52,41 @@ function getCompanyIdsFromBrowser(hash) {
     return cids || [];
 }
 
+const errorHandlerRegistry = registry.category("error_handlers");
+async function accessErrorHandler(env, error, originalError) {
+    const router = env.services.router;
+    const hash = router.current.hash;
+    if (!hash._company_switching) {
+        return false;
+    }
+    if (originalError?.exceptionName === "odoo.exceptions.AccessError") {
+        const { model, id, view_type, menu_id } = hash;
+        if (!model || !id || view_type !== "form") {
+            return false;
+        }
+        router.pushState({ view_type: undefined });
+        if (menu_id) {
+            router.pushState({ menu_id }, { lock: true });
```

Comment:
what about the action_id? You could be in a menu, but not in the action linked to the menu.